docs/source/daml/reference/contract-keys.rst

```diff
@@ -0,0 +1,95 @@
+
+It's best to use simple types for your keys like ``Text`` or ``Integer``, rather than a list or more complex type.
+
+Specifying maintainers
+**********************
+
+If you specify a contract key for a template, you must also specify a ``maintainer`` or maintainers, in a similar way to specifying signatories or observers. Maintainers are the parties that know about all of the keys that they are party to, and are used by the engine to guarantee uniqueness of contract keys.  The maintainers **must** be signatories or observers of the contract.
+
+Keys are unique only to their maintainers. For example, say you have a key that you're using as the identifer for a ``BankAccount`` contract. You might have ``key (bank, accountId) : (Party, Text)``. When you create a new bank account, the contract key ensures that no-one else can have an account with the same ``accountID`` at that bank. But that doesn't apply to other banks: for a contract with a different bank as maintainer, you could happily re-use that ``accountID``.
+
+When you're writing DAML models, the maintainers only matter since they affect authorization -- much like signatories and observers. You don't need to do anything to "maintain" the keys.
+
+Checking of the keys is done automatically at execution time, by the DAML exeuction engine: if someone tries to create a new contract that duplicates an existing contract key, the execution engine will cause that creation to fail. 
+
+Contract keys functions
+***********************
+
+Contract keys introduce several new functions.
+
+``fetchByKey``
+==============
+
+``fetchedContract <- fetchByKey @ContractType contractKey``
+
+Use ``fetchByKey`` to fetch the ID and data of the contract with the specified key. It is an alternative to the currently-used ``fetch``.
+
+It returns a tuple of the ID and the contract object (containing all its data). 
+
+You need authorization from **at least one** of the maintainers to run ``fetchByKey``. A maintainer can authorize by being a signatory, or by submitting the command/being a controller for the choice.
+
+``fetchByKey`` fails and aborts the transaction if:
+
+- you don't have sufficient authorization
+- you're not a stakeholder of the contract you're trying to fetch
+
+This means that if it fails, it doesn't guarantee that a contract with that key doesn't exist, just that you can't see one.
+
+Because the type is ambiguous, when calling you need to specify what you expect with ``@`` and the type, for example ``@MyTemplateType``.
+
+``lookupByKey``
+===============
+
+``contractId <- lookupByKey @ContractType contractKey``
+
+Use ``lookupByKey`` to check whether a contract with the specified key exists. If it does exist, ``lookupByKey`` returns the ``ContractId`` of the contract; otherwise, it returns ``None``.
+
+You need authorization from **all** of the maintainers to run ``lookupByKey``, and it can only be submitted by one of the maintainers.
```

Comment:
Authorization is a local thing within a transaction because it can be delegated from the signatories of a contract into the choices of that contract. The condition that a transaction containing a `lookupByKey` needs to be submitted by one of the maintainers is a global condition. (Is that actually true, @bitonic?) I would split that sentence definitely into two sentences. Maybe like:
```suggestion
You need authorization from **all** of the maintainers to run ``lookupByKey``. A transaction containing a ``lookupByKey`` needs to be submitted by one of the maintainers.
```

Reply:
I'll wait on the answer from @bitonic about whether this is actually right before I add this in :)